Switch WriterProperties to WriterPropertiesFactory trait

corwinjoy opened the pull request:
For review

roeap committed:
refactor: simplify file view iterators

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

roeap committed:
feat: use kernel file skipping for getting files by partition

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

roeap committed:
refactor: move file (action) apis async/stream

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

roeap committed:
refactor: update method invocations to prepare migrations

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

roeap committed:
refactor: use replace EagerSnapshot with LogDataHandler in conflict checks

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

rtyler committed:
chore: remove unnecessary parse_scalar function

PrimitiveType.parse_scalar in kernel does the necessary null handling

Signed-off-by: R. Tyler Croy <rtyler@brokenco.de>

roeap committed:
docs: add some RustDocs

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

roeap committed:
test: add some test for logical file view

Signed-off-by: Robert Pack <robstar.pack@gmail.com>

corwinjoy committed:
Add initial definition for WriterPropertiesFactory

corwinjoy committed:
Switch raw WriterProperties to WriterPropertiesFactory

corwinjoy committed:
Add prototype KMS for WriterPropertiesFactory

corwinjoy committed:
Updated WriterPropertiesBuilder via AI

corwinjoy committed:
cargo fmt

adamreeve commented on crates/core/src/table/table_parquet_options.rs:
```diff
@@ -23,6 +32,34 @@ pub fn build_writer_properties(
 }
 
+#[cfg(feature = "datafusion")]
+pub fn build_writer_properties_factory(
```
Do we want this to handle the case where an encryption factory id is set in the `TableParquetOptions` somehow? Or maybe the `TableEncryption`  type could be integrated in the table properties when those are refactored and then used to create a `KMSWriterPropertiesFactory`? I guess it's not yet clear to me how everything will fit together end-to-end when using KMS encryption.

corwinjoy replied:
Yeah, as mentioned, I think the idea is to create a DeltaTableOptions trait. I would envision it as having two methods:
1. Generate datafusion `TableOptions`. (for reading) - we may want to split this more.
2. Generate a `WriterPropertiesFactory`.

#2 is where it could generate a `KMSWriterPropertiesFactory`. I plan to have a go at adding this over the next day and then you can see if that is enough to generate a workable KMS or if something is missing.

adamreeve reviewed (commented):
This is looking good thanks Corwin, I've left a few comments and questions.

fvaleye committed:
fix(pandas): implement-automatic-conversion-for-pandas-null-types

- implement the conversion of object dtype columns (null) when we found the corresponding type existing in the schema
- works recursively for nested types

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

corwinjoy commented on crates/core/src/operations/optimize.rs:
```diff
@@ -333,18 +336,12 @@ impl<'a> std::future::IntoFuture for OptimizeBuilder<'a> {
 
-            let writer_properties = this.writer_properties.unwrap_or_else(|| {
-                WriterProperties::builder()
-                    .set_compression(Compression::ZSTD(ZstdLevel::try_new(4).unwrap()))
```
It's tough. The standard elsewhere is SNAPPY by default and the code is inconsistent.

corwinjoy committed:
Refactor table_parquet_options functions to further isolate factory construction and standardize defaults.

corwinjoy committed:
cargo fmt

corwinjoy committed:
Add tracing to confirm that all files pass through create_writer_properties

fvaleye committed:
feat(pandas): use field access using arro3 on type conversion

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

adamreeve commented on crates/core/src/operations/write/writer.rs:
```diff
@@ -283,16 +289,17 @@ impl PartitionWriterConfig {
-        let writer_properties = writer_properties.unwrap_or_else(|| {
-            WriterProperties::builder()
+        let writer_properties_factory = writer_properties_factory.unwrap_or_else(|| {
+            let wp = WriterProperties::builder()
```
> Sometimes the default compression is SNAPPY, sometimes it is ZSTD

I think that might be intentional, and snappy is chosen for its speed for normal writes, while zstd is chosen for superior compression during compaction.

rtyler committed:
fix: enabling correctly pulling partition values out of column mapped tables

Juggling this unit test took a *long* time but found a few other places
that needed to use the right names in the process :weary:

Signed-off-by: R. Tyler Croy <rtyler@brokenco.de>

rtyler committed:
chore: remove unused dependencies

Signed-off-by: R. Tyler Croy <rtyler@brokenco.de>

fvaleye committed:
fix(format): fix formatting in Python for conversion file

Signed-off-by: Florian Valeye <florian.valeye@gmail.com>

rtyler committed:
Update crates/core/src/kernel/snapshot/mod.rs

Co-authored-by: Ion Koutsouris <15728914+ion-elgreco@users.noreply.github.com>

corwinjoy committed:
Change build_writer_properties_tpo to public.

corwinjoy committed:
Change default compression to SNAPPY

corwinjoy committed:
Merge branch 'main' into crypto_config3

# Conflicts:
#	crates/core/src/operations/optimize.rs
#	crates/core/tests/command_optimize.rs

corwinjoy committed:
cargo fmt

corwinjoy committed:
fix incorrect arguments from merge
